### PR TITLE
Fix empty detach keys being interpreted as default detach keys

### LINKED ISSUE #28193: podman --remote exec ignores empty detach keys
I've confirmed the fix in 5.8.0 for https://github.com/containers/podman/issues/27414 but it only fixes the issue for podman run, not podman exec. In the latter case, an empty string is treated as nothing and the default "C-p,C-q" is used instead of disabling the detach sequence.

### PR DESCRIPTION
**Pull request by Ibrahim2750mi**
Fixes #28193
As mentioned in the issue and the linked PR #28405, the problem was string doesn't hold necessary information to conclude if detached keys were passed as `''` or omitted. So, the obvious fix was to change the `input.DetachKeys` to *string.

If default key is omitted, in `exec.go` we assign `libpodConfig.DetachKeys` to `&config.DefaultDetachKeys`.

### CODE REVIEW

**Review by mheon — APPROVED** [review]
LGTM

File: pkg/api/handlers/types.go [hunk 1]
```diff
@@ -237,6 +237,7 @@ type HistoryResponse struct {
 
 type ExecCreateConfig struct {
 	dockerContainer.ExecCreateRequest
+	DetachKeys *string `json:"DetachKeys"`
```
Comment (Luap99) [inline pkg/api/handlers/types.go:248]: That is a breaking change for pkg/bindings users, types there must be stable as they are used by external users.

The fact that you need to switch it in pkg/domain/infra/tunnel/containers.go shows that.

This needs a second type that is only used on the api side since ExecCreateConfig is currently exposed in pkg/bindings.

**Review by Luap99 — CHANGES_REQUESTED** [review]

File: pkg/api/handlers/types.go (hunk 1 above)
Comment (Ibrahim2750mi) [reply pkg/api/handlers/types.go:248]: Should I add the new type in `types.go` only? or I can add it in `exec.go`? If we add it in `types.go`, are we deprecating the older type in future?

Comment (Ibrahim2750mi) [reply pkg/api/handlers/types.go:248]: @Luap99 any input?

Comment (Ibrahim2750mi) [reply pkg/api/handlers/types.go:248]: @mheon ?

Comment (Luap99) [reply pkg/api/handlers/types.go:248]: we can add a second type in that file and deprecate the old one slowly yes.

Comment (Ibrahim2750mi) [reply pkg/api/handlers/types.go:248]: addressed by https://github.com/podman-container-tools/podman/pull/29507/commits/29332e02358f132a862bd47a11e6eb47d363c0dd

File: pkg/api/handlers/compat/exec.go [hunk 2]
```diff
@@ -53,9 +54,13 @@ func ExecCreateHandler(w http.ResponseWriter, r *http.Request) {
 	libpodConfig.AttachStdin = input.AttachStdin
 	libpodConfig.AttachStderr = input.AttachStderr
 	libpodConfig.AttachStdout = input.AttachStdout
-	if input.DetachKeys != "" {
-		libpodConfig.DetachKeys = &input.DetachKeys
+
+	libpodConfig.DetachKeys = input.DetachKeys
+
+	if libpodConfig.DetachKeys == nil {
+		libpodConfig.DetachKeys = &config.DefaultDetachKeys
```
Comment (Honny1) [inline pkg/api/handlers/compat/exec.go]: The current change ignores any custom detach key configuration the user may have set in `containers.conf`.

```suggestion
		libpodConfig.DetachKeys = &runtimeConfig.Engine.DetachKeys
```

Comment (Honny1) [reply pkg/api/handlers/compat/exec.go]: Maybe a `nil` check should be done where `runtimeConfig.Engine.DetachKeys` is populated.

File: pkg/api/handlers/types.go [hunk 3]
```diff
@@ -235,10 +235,19 @@ type HistoryResponse struct {
 	Comment   string
 }
 
+
+// #29507 - DeprecationWarning: will be replaced by `ExecCreateConfigJSON` in future.
```
Comment (Honny1) [inline pkg/api/handlers/types.go]: This doesn't follow Go conventions. It should look like this: `// Deprecated: use ExecCreateConfigJSON instead.`

File: test/apiv2/20-containers.at [hunk 4]
```diff
@@ -948,6 +948,41 @@ t POST containers/create?platform=linux/aarch64 \
   Image=$IMAGE \
   404
 
+# 28193: podman --remote exec ignores empty detach keys (explicitly disable detach).
```
Comment (Honny1) [inline test/apiv2/20-containers.at:951]: Can you please create e2e or system test for the `podman exec --detach-keys ""`.

File: pkg/api/handlers/types.go [hunk 5]
```diff
@@ -235,10 +235,19 @@ type HistoryResponse struct {
...
 type ExecCreateConfig struct {
 	dockerContainer.ExecCreateRequest
 }
 
+// #28193: added second type for JSON unmarshling, to prevent breaking change in `pkg/bindings`.
+type ExecCreateConfigJSON struct {
+	dockerContainer.ExecCreateRequest
+	DetachKeys *string `json:"DetachKeys"`
```
Comment (Honny1) [inline pkg/api/handlers/types.go]: I would change `ExecCreateConfig` struct to: 

```
type ExecCreateConfig struct {
    dockerContainer.ExecCreateRequest
    DetachKeys string `json:"DetachKeys,omitempty"`
}
```

This changes nothing about the Go type. `DetachKeys` is still a string, external code sets `config.DetachKeys = "ctrl-a"` still compiles and works. 

The only behavioral change is that when `DetachKeys` is `""`, it's omitted from JSON instead of sent as `""`. On the server, the `ExecCreateConfigJSON` (Personally, I would like to see a better name, for example, something like `ExecCreateConfigBody` or `ExecCreateConfigRequestBody`) type then sees `nil` (field absent) and correctly falls back to the default.

**Review by Honny1 — CHANGES_REQUESTED** [review]
I am not a fan of depreciation, but I have a different idea for how to do that. Also, I would like to see integration tests.

File: pkg/api/handlers/types.go (hunk 5 above)
Comment (Luap99) [reply pkg/api/handlers/types.go]: > The only behavioral change is that when DetachKeys is "", it's omitted from JSON instead of sent as "". On the server, the ExecCreateConfigJSON (Personally, I would like to see a better name, for example, something like ExecCreateConfigBody or ExecCreateConfigRequestBody) type then sees nil (field absent) and correctly falls back to the default.

Which is exactly what we do not want. `podman exec --detach-keys ""` must send the empty string over the write so the server will not pick the default but treat that as no detaching.

File: test/apiv2/20-containers.at (hunk 4 above)
Comment (Luap99) [reply test/apiv2/20-containers.at:951]: the problem with e2e tests is that at the moment one cannot safely detach form scripts

https://github.com/podman-container-tools/podman/pull/25083 has a long history with many small corner cases as problems which I never had the time to get back to

File: pkg/api/handlers/types.go (hunk 5 above)
Comment (Honny1) [reply pkg/api/handlers/types.go]: Ah, missed that case.

File: pkg/api/handlers/types.go (hunk 3 above)
Comment (Honny1) [inline pkg/api/handlers/types.go]: This doesn't appear to be a proper Go depreciation notice.

Comment (Ibrahim2750mi) [reply pkg/api/handlers/types.go]: addressed by https://github.com/podman-container-tools/podman/pull/29507/commits/9a711e96cb290fa8dc4a353e8878defbe56e31cf

File: pkg/api/handlers/compat/exec.go (hunk 2 above)
Comment (Ibrahim2750mi) [reply pkg/api/handlers/compat/exec.go]: addressed by https://github.com/podman-container-tools/podman/pull/29507/commits/9a711e96cb290fa8dc4a353e8878defbe56e31cf

File: pkg/api/handlers/types.go [hunk 6]
```diff
@@ -235,10 +235,20 @@ type HistoryResponse struct {
...
+// Deprecated: use `ExecCreateConfigBody` instead.
 type ExecCreateConfig struct {
 	dockerContainer.ExecCreateRequest
 }
 
+// #28193: added second type for JSON unmarshling, to prevent breaking change in `pkg/bindings`.
+type ExecCreateConfigBody struct {
+	dockerContainer.ExecCreateRequest
+	DetachKeys *string `json:"DetachKeys"`
+
```
Comment (Honny1) [inline pkg/api/handlers/types.go]: Nonblocking: This white space is not needed.

File: pkg/api/handlers/compat/exec.go [hunk 7]
```diff
@@ -53,9 +53,9 @@ func ExecCreateHandler(w http.ResponseWriter, r *http.Request) {
 	libpodConfig.AttachStdin = input.AttachStdin
 	libpodConfig.AttachStderr = input.AttachStderr
 	libpodConfig.AttachStdout = input.AttachStdout
-	if input.DetachKeys != "" {
-		libpodConfig.DetachKeys = &input.DetachKeys
-	}
+
```
Comment (Honny1) [inline pkg/api/handlers/compat/exec.go]: Nonblocking: This white space is not needed.

File: pkg/api/handlers/compat/exec.go [hunk 8]
```diff
@@ -53,9 +53,9 @@ func ExecCreateHandler(w http.ResponseWriter, r *http.Request) {
...
+	libpodConfig.DetachKeys = input.DetachKeys
+
```
Comment (Honny1) [inline pkg/api/handlers/compat/exec.go]: Nonblocking: This white space is not needed.

**Review by Honny1 — APPROVED** [review]
LGTM, just nonblocking comments.

PTAL @podman-container-tools/podman-maintainers @podman-container-tools/podman-reviewers

File: pkg/api/handlers/compat/exec.go (hunk 8 above)
Comment (Ibrahim2750mi) [reply pkg/api/handlers/compat/exec.go]: addressed by https://github.com/podman-container-tools/podman/pull/29507/commits/97742a7ab2e5f6c790aa9a6a8e2c4d9b803f47ec

File: pkg/api/handlers/compat/exec.go (hunk 7 above)
Comment (Ibrahim2750mi) [reply pkg/api/handlers/compat/exec.go]: addressed by https://github.com/podman-container-tools/podman/pull/29507/commits/97742a7ab2e5f6c790aa9a6a8e2c4d9b803f47ec

File: pkg/api/handlers/types.go (hunk 6 above)
Comment (Ibrahim2750mi) [reply pkg/api/handlers/types.go]: addressed by https://github.com/podman-container-tools/podman/pull/29507/commits/97742a7ab2e5f6c790aa9a6a8e2c4d9b803f47ec

**Review by Honny1 — APPROVED** [review]
LGTM

File: pkg/api/handlers/exec_test.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	"encoding/json"
+	"testing"
+
+	"go.podman.io/podman/v6/pkg/api/handlers"
+)
+
+// TestExecCreateConfigDetachKeys verifies that DetachKeys *string correctly
+// distinguishes "field absent" (nil) from "explicitly set to empty string" (*"").
+// #28193 - podman --remote exec ignores empty detach keys
+func TestExecCreateConfigDetachKeysUnmarshal(t *testing.T) {
```
Comment (Honny1) [inline pkg/api/handlers/exec_test.go]: Nonblocking: I am not sure if having this type of test is beneficial. Imho basically tests `json.Unmarshal`, whytch is not need to be tested by us.

**Review by Honny1 — COMMENTED** [review]
I would ingore linter issues.

File: pkg/api/handlers/exec_test.go (hunk 9 above)
Comment (Ibrahim2750mi) [reply pkg/api/handlers/exec_test.go]: @Luap99 thoughts?

Comment (Ibrahim2750mi) [reply pkg/api/handlers/exec_test.go]: @mheon ?

Comment (mheon) [reply pkg/api/handlers/exec_test.go]: Yeah, agree with @Honny1 - test can be dropped

Comment (Ibrahim2750mi) [reply pkg/api/handlers/exec_test.go]: So, I should drop this file? @Honny1 @Luap99

Comment (Honny1) [reply pkg/api/handlers/exec_test.go]: Yes

Comment (Ibrahim2750mi) [reply pkg/api/handlers/exec_test.go]: addressed by https://github.com/podman-container-tools/podman/pull/29507/commits/e4c484b37ac5e29c3b97f0a17128e7d7b7a9233f

**Comment by Honny1** [comment]
@Ibrahim2750mi Any update on this? 

```
+ /home/ubuntu/golangci-lint-2.12.2-linux-amd64/golangci-lint run --build-tags=apparmor,seccomp,selinux
Error: pkg/bindings/containers/exec.go:24:63: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
func ExecCreate(ctx context.Context, nameOrID string, config *handlers.ExecCreateConfig) (string, error) {
                                                              ^
Error: pkg/bindings/test/exec_test.go:38:21: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
		execConfig := new(handlers.ExecCreateConfig)
		                  ^
Error: pkg/bindings/test/exec_test.go:67:21: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
		execConfig := new(handlers.ExecCreateConfig)
		                  ^
Error: pkg/bindings/test/exec_test.go:74:21: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
		execConfig := new(handlers.ExecCreateConfig)
		                  ^
Error: pkg/domain/infra/tunnel/containers.go:604:52: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
func makeExecConfig(options entities.ExecOptions) *handlers.ExecCreateConfig {
                                                   ^
Error: pkg/domain/infra/tunnel/containers.go:610:22: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
	createConfig := new(handlers.ExecCreateConfig)
	                    ^
6 issues:
* staticcheck: 6
```

**Comment by Honny1** [comment]
@Ibrahim2750mi Any updates on this?

**Comment by Ibrahim2750mi** [comment]
```
+ /home/ubuntu/golangci-lint-2.12.2-linux-amd64/golangci-lint run --build-tags=apparmor,seccomp,selinux
Error: pkg/bindings/containers/exec.go:24:63: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
func ExecCreate(ctx context.Context, nameOrID string, config *handlers.ExecCreateConfig) (string, error) {
                                                              ^
Error: pkg/bindings/test/exec_test.go:38:21: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
		execConfig := new(handlers.ExecCreateConfig)
		                  ^
Error: pkg/bindings/test/exec_test.go:67:21: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
		execConfig := new(handlers.ExecCreateConfig)
		                  ^
Error: pkg/bindings/test/exec_test.go:74:21: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
		execConfig := new(handlers.ExecCreateConfig)
		                  ^
Error: pkg/domain/infra/tunnel/containers.go:604:52: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
func makeExecConfig(options entities.ExecOptions) *handlers.ExecCreateConfig {
                                                   ^
Error: pkg/domain/infra/tunnel/containers.go:610:22: SA1019: handlers.ExecCreateConfig is deprecated: use `ExecCreateConfigBody` instead. (staticcheck)
	createConfig := new(handlers.ExecCreateConfig)
	                    ^
6 issues:
* staticcheck: 6
```

@Honny1  Out of the 6 linter issues, which issues are supposed to be fixed? 2, 3, 4, 5 and 6? They seem non-breaking to me. Additionally what to do about 1?
(numbered in order from top to bottom, top -> 1)

**Comment by Honny1** [comment]
@Ibrahim2750mi I would ignore these findings with //nolint:staticcheck // Ignore SA1019: ...REASON.... Fill in the reason with something like: "deprecated, but retained for backward compatibility."
